### PR TITLE
Don't return duplicate rows in the customer addresses report

### PR DESCRIPTION
**Pull request by cillian**
#### What? Why?

- Closes https://github.com/openfoodfoundation/openfoodnetwork/issues/10706

https://github.com/openfoodfoundation/openfoodnetwork/assets/101088/c64c79f4-aebb-4653-992f-76412741ce03

#### What should we test?

See steps to reproduce in https://github.com/openfoodfoundation/openfoodnetwork/issues/10706

#### Release notes

Changelog Category: User facing changes

### CODE REVIEW

File: spec/lib/reports/customers_report_spec.rb [hunk 1]
```diff
@@ -140,30 +140,62 @@ module Customers
...
+                                                       o1.email, a.phone, d.name,
+                                                       [d.address.address1, d.address.address2, d.address.city].join(" "),
+                                                       o1.shipping_method.name
+                                                     ], [
+                                                       a.firstname, a.lastname,
+                                                       [a.address1, a.address2, a.city].join(" "),
+                                                       o2.email, a.phone, d.name,
+                                                       [d.address.address1, d.address.address2, d.address.city].join(" "),
+                                                       sm2.name
+                                                     ]])
```
Comment (rioug) [inline spec/lib/reports/customers_report_spec.rb:173]: I am nitpicking a little bit, but could you reformat this as you did below ? much more readable !

**Review by rioug — APPROVED** [review]
It looks good ! Thanks for that. 
I did leave a comment around formatting, but it shouldn't block merging.

**Comment by cillian** [comment]
@rioug Thanks for the review, that formatting is fixed now.

**Review by rioug — APPROVED** [review]

**Review by dacook — COMMENTED** [review]
Interesting way to solve the problem. I would have expected to solve it in within the report itself.

I think in [lib/reporting/reports/customers/addresses.rb](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/lib/reporting/reports/customers/addresses.rb#L8) we could dynamically choose the contents of `group_by` according to the selected columns. So if shipping_method is present, group by it, but if not we don't.

But I'm not sure how to do that exactly. This reports framework is complex.
So let's do this 👍 

I do have a suggestion for slightly more efficient code, so will apply that first.

**Review by dacook — APPROVED** [review]

**Comment by dacook** [comment]
@rioug would you mind reviewing my last commit?

**Review by rioug — APPROVED** [review]

**Comment by drummer83** [comment]
Hi @cillian,
Thanks for working on this! It's working great for the Customer Addresses report. As you can see, before your PR we had duplicates in the list if not all columns are shown (shipping method is hidden). After your PR those duplicates are removed. Awesome!

**Before your PR**
image

**After your PR**
image

I have also tested that the same is true for the exported files. :+1: 

One thing which was not very clear in the issue description, I have to admit, is that we have the same behavior in the Customers Mailing List report. However it seems that [we will be removing that report soon](https://github.com/openfoodfoundation/openfoodnetwork/issues/11155#issue-1787540623) - so no additional action required here.

Thanks again for this great fix! Merging! :rocket:

**Pull request merged**